zerver/tests/test_custom_profile_data.py

```diff
@@ -888,6 +964,94 @@ def test_default_external_account_type_field(self) -> None:
 
+    def test_update_with_editable_by_user(self) -> None:
+        iago = self.example_user("iago")
+        aaron = self.example_user("aaron")
+
+        self.login("iago")
+
+        # Create field with editable_by_user = false
+        data: Dict[str, Any] = {}
+        data["name"] = "Name "
+        data["field_type"] = CustomProfileField.SHORT_TEXT
+        data["editable_by_user"] = "false"
+        result = self.client_post("/json/realm/profile_fields", info=data)
+        result_json = self.assert_json_success(result)
+        restricted_field_id = result_json["id"]
+
+        field_data = [
+            {
+                "id": restricted_field_id,
+                "value": "test",
+            }
+        ]
+
+        # Admins can always change their own fields
+        for field_dict in iago.profile_data():
+            if field_dict["id"] == restricted_field_id:
+                self.assertEqual(field_dict["value"], None)
+        result = self.client_patch(
+            "/json/users/me/profile_data", {"data": orjson.dumps(field_data).decode()}
+        )
+        self.assert_json_success(result)
+        for field_dict in iago.profile_data():
+            if field_dict["id"] == restricted_field_id:
+                self.assertEqual(field_dict["value"], "test")
+
+        # Admins can always change fields of others
+        for field_dict in aaron.profile_data():
+            if field_dict["id"] == restricted_field_id:
+                self.assertEqual(field_dict["value"], None)
+        result = self.client_patch(
+            f"/json/users/{aaron.id}", {"profile_data": orjson.dumps(field_data).decode()}
+        )
+        self.assert_json_success(result)
+        for field_dict in aaron.profile_data():
+            if field_dict["id"] == restricted_field_id:
+                self.assertEqual(field_dict["value"], "test")
+
+        # Login as normal user
+        self.logout()
+        self.login_user(aaron)
+
+        # Users cannot update field value when editable_by_user is false
+        for field_dict in aaron.profile_data():
+            if field_dict["id"] == restricted_field_id:
+                self.assertEqual(field_dict["value"], "test")
+        result = self.client_patch(
+            "/json/users/me/profile_data", {"data": orjson.dumps(field_data).decode()}
+        )
+        self.assert_json_error(result, "User does not have permission to update the field.")
+        for field_dict in aaron.profile_data():
+            if field_dict["id"] == restricted_field_id:
+                self.assertEqual(field_dict["value"], "test")
+
+        # Login as admin and change editable_by_user to true
+        self.logout()
+        self.login_user(iago)
+        data = {}
+        data["name"] = "Name "
+        data["editable_by_user"] = "true"
+        result = self.client_patch(f"/json/realm/profile_fields/{restricted_field_id}", info=data)
+        self.assert_json_success(result)
```

Comment:
You can use `self.api_patch` to do the administrator operations without logging in as them; that often results in more readable tests.

Reply:
Modified to use `self.api_patch` for admin actions, still using `self.client_patch` for user changes though. 

Since both `/api/v1/*` and /json/*` ultimately call the same function, i assume we don't need to test both paths?

Reply:
Correct!